### PR TITLE
Font Library: Make notices more consistent

### PR DESCRIPTION
**Pull request by mikachan**
## What?

This PR attempts to improve the way notices are handled in the Font Library modal, by passing the `notice` up to the `TabPanelLayout` container component.

This PR also includes a change to make the notices manually dismissible, in line with this comment: https://github.com/WordPress/gutenberg/pull/55975#issuecomment-1820584893

Please note that local-fonts.js has been renamed upload-fonts.js, as local-fonts.js was just a container with a spacer that imported upload-fonts.js. The spacer is no longer needed, as it looks like it was only there to provide spacing between the notice and the upload component, and this is now handled in `TabPanelLayout`. See further information here: https://github.com/WordPress/gutenberg/pull/58180#discussion_r1464914306.

## Why?

This makes all the notices in the modal appear in the same place and means that they are all handled via `TabPanelLayout`, rather than separately by each component. It also makes the notices easier to update as there is less code to change.

## How?

- `Notice` component has been removed from `FontCollection`, `InstalledFonts`, and `LocalFonts`
- `Notice` component has been added to `TabPanelLayout` component
- Notice logic is still the same for `FontCollection`, `InstalledFonts`, and `LocalFonts`, but the state has been moved from the local state of each component to the global Font Library state

## Testing Instructions

1. Open the Font Library modal
2. Trigger various notices in all the tabs (Library, Upload, Font Collection)
3. Ensure all notices appear correctly and in the same place

The easiest notice to show is the success notice when a new font is installed. Alternatively, force a notice to show in the `TabPanelLayout` component, for example by adding this just before the return on line 29:

```
notice = {
	type: 'success',
	message: 'This is a test notice',
};
```

### Library Tab
| Before | After |
| ------ | ---- |
| image | image |

### Upload Tab
| Before | After |
| ------ | ---- |
| image | image |

### Font Collection Tab
| Before | After |
| ------ | ---- |
| image | image |

### CODE REVIEW

File: packages/edit-site/src/components/global-styles/font-library-modal/font-demo.js [hunk 1]
```diff
@@ -62,7 +62,7 @@ function FontFaceDemo( { customPreviewUrl, fontFace, text, style = {} } ) {
 			}
 		};
 		loadAsset();
-	}, [ fontFace, isIntersecting, loadFontFaceAsset ] );
+	}, [ fontFace, isIntersecting, loadFontFaceAsset, isPreviewImage ] );
```
Comment (mikachan) [inline packages/edit-site/src/components/global-styles/font-library-modal/font-demo.js:65]: This is unrelated, I just noticed it needed updating.

File: packages/edit-site/src/components/global-styles/font-library-modal/upload-fonts.js [hunk 2]
```diff
@@ -1,19 +1,214 @@
 /**
```
Comment (mikachan) [inline packages/edit-site/src/components/global-styles/font-library-modal/upload-fonts.js:1]: This file has been renamed from local-fonts.js to upload-fonts.js, as local-fonts.js was just a container with a spacer that imported upload-fonts.js. The spacer is no longer needed, as it looks like it was only there to provide spacing between the notice and the upload component, and this is now handled in `TabPanelLayout`.

Comment (pbking) [reply packages/edit-site/src/components/global-styles/font-library-modal/upload-fonts.js:1]: It took me a minute to understand that all of the changes were from the removed 'local-fonts.js`.  Thank you for explaining what happened. :)

**Comment by pbking** [comment]
This has tested well for me.  I've confirmed that success and error messages are both shown as expected, the position is where I expected, the message does not go away after a time.

Two things to note: the alert DOES go away when a font-collection tab is selected (intentional I believe) but does NOT go away when the Upload or Library tabs are selected.  I believe it should ALWAYS remain until dismissed or ANY tab change should dismiss the message for consistency.

**Comment by mikachan** [comment]
Thanks @pbking!

> I believe it should ALWAYS remain until dismissed or ANY tab change should dismiss the message for consistency.

Makes sense - I've added a [commit](https://github.com/WordPress/gutenberg/pull/58180/commits/2bf44c29fff28b50c6a855cd3384f1a4253f813e) that resets the notice whenever a new tab is selected. Hopefully this addresses this. Ready for another review!

**Review by creativecoder — COMMENTED** [review]
The code changes look reasonable to me and sharing the notice component is a nice improvement in the code organization for the font library.

Something about the notice being above the header within the tab makes the content jump when the notice is inserted more disturbing to me. This has me wondering if this is the wrong UI pattern to use here, or maybe there's a better place to put the notice. 🤔 

font-library-notice

**Comment by mikachan** [comment]
Thanks @creativecoder!

> This has me wondering if this is the wrong UI pattern to use here, or maybe there's a better place to put the notice. 🤔

I'm also unsure about the notice placement and the layout shift, I'm going to take a look at some options in a separate PR.

I've added another commit so that the notice is also reset on each action: https://github.com/WordPress/gutenberg/pull/58180/commits/e7fb1e85167d5a5894046b3afadca9a77d971944.

**Review by matiasbenedetto — COMMENTED** [review]
When the list of font faces require scroll the notification won't be visible.
Should we make the notice float above the elements?


https://github.com/WordPress/gutenberg/assets/1310626/b8d36116-d1ab-4d6f-9b6e-f0a0aab12cf4

**Comment by matiasbenedetto** [comment]
In this case where one of the font collections fail to load the data: should the notification be removed if you switch to another tab?



https://github.com/WordPress/gutenberg/assets/1310626/46684099-c688-447f-a32d-1fcfc47f8342

**Comment by mikachan** [comment]
Thanks @matiasbenedetto!

> Should we make the notice float above the elements?

I was going to explore that in a different PR as it's a further deviation from the original design, but I can continue in this PR if you think that's best.

> In this case where one of the font collections fail to load the data: should the notification be removed if you switch to another tab?

This should be handled by the tabs `onSelect` function. It looks like `setNotice` is causing the collection tab to re-render if there's an error, this should be fixed with https://github.com/WordPress/gutenberg/pull/58180/commits/9705a486103d77b2f96cc4bdd2ad92db1de9ea00.

**Comment by matiasbenedetto** [comment]
> I was going to explore that in a different PR as it's a further deviation from the original design.

Yep, it's a good idea to implement it in a follow-up PR.

**Review by matiasbenedetto — APPROVED** [review]
This is looking and working well on my tests.
LGTM.

**Comment by mikachan** [comment]
> Something about the notice being above the header within the tab makes the content jump when the notice is inserted more disturbing to me.

@creativecoder I think I have a potential improvement for this in https://github.com/WordPress/gutenberg/pull/58180/commits/ccf7f40cdbc61a40d0fa3b56ae09c5a7752be440. This moves the notice below the font collection header section, but it's still in the same place for the other tabs:

| Library | Upload | Font Collection |
| ---- | ----  |--- |
| image | image | image |

I think this is most similar to the current implementation, and then we can explore further design options in a separate PR if needed.

**Comment by mikachan** [comment]
Sorry @matiasbenedetto, I posted an update just as you commented 🙈 It's only a small change and I think it's an improvement.

**Comment by matiasbenedetto** [comment]
> Sorry @matiasbenedetto, I posted an update just as you commented 🙈 It's only a small change and I think it's an improvement.

It looks better, nice improvement :+1:

File: packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js [hunk 3]
```diff
@@ -77,36 +75,27 @@ function FontCollection( { slug } ) {
 				await getFontCollection( slug );
 				resetFilters();
 			} catch ( e ) {
-				setNotice( {
-					type: 'error',
-					message: e?.message,
-					duration: 0, // Don't auto-hide.
-				} );
+				if ( ! notice ) {
```
Comment (pbking) [inline packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js:78]: Do we have to check for notice?  I think we could remove this line.

**Review by pbking — APPROVED** [review]
Left a tiny comment about maybe removing one line.  Code looks great and it works exactly like I would hope it to.  Nicely done.

File: packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js (hunk 3 above)
Comment (mikachan) [reply packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js:78]: I added this as the `fetchFontCollection` function keeps re-setting the notice on failure if there is an error with the font collection (e.g. something wrong with the JSON file), and it was causing an infinite loop of setting the error notice, I think because `setNotice` is a dependency of the `useEffect` where this function is called, so the notice was causing a re-render. This line fixed the issue (original discussion [here](https://github.com/WordPress/gutenberg/pull/58180#issuecomment-1910358720)). (This is happening on trunk, so I guess this isn't 100% related to this PR.)

File: packages/edit-site/src/components/global-styles/font-library-modal/upload-fonts.js (hunk 2 above)
Comment (afercia) [reply packages/edit-site/src/components/global-styles/font-library-modal/upload-fonts.js:1]: A few pending PRs have now conflicts because of this file renaming. It took me a while to research and understand what happened. Thank you for explaining.it. Mentioning it in the PR description would be even more appreciated, thanks,

Comment (mikachan) [reply packages/edit-site/src/components/global-styles/font-library-modal/upload-fonts.js:1]: Updated the PR description to include this info as well, thanks!

**Comment by cbravobernal** [comment]
I just cherry-picked this PR to the release/17.6 branch to get it included in the next release: 6715cf5087
